src/compiler/crystal/semantic/top_level_visitor.cr

```diff
@@ -41,13 +41,29 @@ class Crystal::TopLevelVisitor < Crystal::SemanticVisitor
 
   @last_doc : String?
 
+  # special types recognized for `@[Primitive]`
+  private enum PrimitiveType
+    ReferenceStorageType
+  end
+
   def visit(node : ClassDef)
     check_outside_exp node, "declare class"
 
     scope, name, type = lookup_type_def(node)
 
     annotations = read_annotations
 
+    special_type = nil
+    process_annotations(annotations) do |annotation_type, ann|
+      case annotation_type
+      when @program.primitive_annotation
+        arg = ann.args.first.as(SymbolLiteral)
+        unless special_type = PrimitiveType.parse?(arg.value)
```

Comment:
**nitpick(style):** I found it hard to recognize that this line assigns to `special_type`. I was looking for an assignment and didn't see this at first because it starts with an `unless`.
I'd suggest to split this into two lines to improve readability.
```suggestion
        special_type = PrimitiveType.parse?(arg.value)
        unless special_type
```
A trailing `unless` would also be possible.